File: python/example5_1.py
# Example 5.1
# Condition Number and Correlation Matrix
%cd C:/Course19/ceR/python

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf

data=pd.read_csv('http://web.pdx.edu/~crkl/ceR/data/longley.txt',sep='\s+',nrows=16)

model = smf.ols(formula='EM ~ YEAR + PGNP + GNP + AF',data = data).fit() 
print(model.summary())

#condition number 
model.condition_number
# cm=model.eigenvals
# np.sqrt(max(cm)/min(cm))

#correlation matrix 
data[['EM', 'YEAR', 'PGNP', 'GNP', 'AF']].corr()

# alternatively, using design-matrix to compute the condition number
# pip install patsy
import patsy as pt
y, X = pt.dmatrices('EM ~ YEAR + PGNP + GNP + AF',data = data)
np.linalg.cond(X)

File: R/example5_1.R
# Example 5.1
# Condition Number and Correlation Matrix

setwd("C:/Course17/ceR/R")
data<-read.table("http://web.pdx.edu/~crkl/ceR/data/longley.txt",header=T,nrows=16)

year<-data$YEAR
pgnp<-data$PGNP
gnp<-data$GNP
af<-data$AF
em<-data$EM

results<-lm(em~year+pgnp+gnp+af)
summary(results)

#  Correlation Matrix
y<-cbind(year,pgnp,gnp,af,em)
y1<-cor(y)
y1

# Condition Number
x<-cbind(1,year,pgnp,gnp,af)
x1<-t(x)%*%x
cm<-eigen(x1)$val
cn<-sqrt(max(cm)/min(cm))
cn
kappa(x) # 2-norm condition number 

rm(list=ls())
